std/assembly/encoding.ts

```diff
@@ -0,0 +1,194 @@
+    if (<usize>len > <usize>MAX_BYTELENGTH) throw new RangeError(E_INVALIDLENGTH);
+    var size = <usize>len;
+    var str = ALLOCATE(size);
+    memory.copy(str, changetype<usize>(buf), size);
+    return REGISTER<string>(str);
+  }
+}
+
+/** UTF8 encoding. */
+export namespace UTF8 {
+
+  /** Calculates the length of a string when encoded as an UTF8 buffer. */
+  export function length(str: string, delimited: bool = false): i32 {
+    var strOff = changetype<usize>(str);
+    var strEnd = changetype<usize>(str) + (<usize>str.length << 1);
+    var bufLen = delimited ? 1 : 0;
+    while (strOff < strEnd) {
+      let c = <u32>load<u16>(strOff);
+      if (c < 128) {
+        bufLen += 1; strOff += 2;
+      } else if (c < 2048) {
+        bufLen += 2; strOff += 2;
+      } else {
+        if ((c & 0xFC00) == 0xD800 && strOff + 2 < strEnd) {
+          if ((<u32>load<u16>(strOff, 2) & 0xFC00) == 0xDC00) {
+            strOff += 4; bufLen += 4;
+            continue;
+          }
+        }
+        strOff += 2; bufLen += 3;
+      }
+    }
+    return bufLen;
+  }
+
+  /** Encodes a string as an UTF8 buffer. */
+  export function encode(str: string, delimited: bool = false): ArrayBuffer {
+    var strOff = changetype<usize>(str);
+    var strEnd = changetype<usize>(str) + (<usize>str.length << 1);
+    var buf = ALLOCATE(<usize>length(str, delimited));
+    var bufOff = changetype<usize>(buf);
+    while (strOff < strEnd) {
+      let c1 = <u32>load<u16>(strOff);
+      if (c1 < 128) {
+        store<u8>(bufOff, c1);
+        bufOff += 1; strOff += 2;
+      } else if (c1 < 2048) {
+        store<u8>(bufOff, c1 >> 6      | 192);
+        store<u8>(bufOff, c1      & 63 | 128, 1);
+        bufOff += 2; strOff += 2;
+      } else {
+        if ((c1 & 0xFC00) == 0xD800 && strOff + 2 < strEnd) {
+          let c2 = <u32>load<u16>(strOff, 2);
+          if ((c2 & 0xFC00) == 0xDC00) {
+            c1 = 0x10000 + ((c1 & 0x03FF) << 10) + (c2 & 0x03FF);
+            store<u8>(bufOff, c1 >> 18      | 240);
+            store<u8>(bufOff, c1 >> 12 & 63 | 128, 1);
+            store<u8>(bufOff, c1 >> 6  & 63 | 128, 2);
+            store<u8>(bufOff, c1       & 63 | 128, 3);
+            strOff += 4; bufOff += 4;
+            continue;
+          }
+        }
+        store<u8>(bufOff, c1 >> 12      | 224);
+        store<u8>(bufOff, c1 >> 6  & 63 | 128, 1);
+        store<u8>(bufOff, c1       & 63 | 128, 2);
+        strOff += 2; bufOff += 3;
+      }
+    }
+    assert(strOff == strEnd);
+    if (delimited) store<u8>(bufOff, 0);
+    return REGISTER<ArrayBuffer>(buf);
+  }
+
+  /** Decodes an UTF8 buffer to a string.*/
+  export function decode(buf: ArrayBuffer, delimited: bool = false): string {
+    return delimited
+      ? decodeRawDelimited(changetype<usize>(buf), buf.byteLength)
+      : decodeRaw(changetype<usize>(buf), buf.byteLength);
+  }
+
+  // @ts-ignore: decorator
+  @unsafe
+  export function decodeRaw(buf: usize, len: i32): string {
```

Comment:
It looks like this could be internal because `decode` already provide more general interface. WDYT?

Reply:
The idea is that someone who gets just a `str` and a `str_len` from C has a way to make a `string` from it, as the higher level `decode´ function wants an `ArrayBuffer` which must have a runtime header with payloadLength.

Reply:
So may be makes sense add optional `len` to `decode` as well?

Reply:
That'd look like `UTF8.decode(changetype<ArrayBuffer>(str), false, str_len)` then. Not sure, feels somewhat wrong.

Reply:
`UTF8.decode(changetype<ArrayBuffer>(str), 0, true)` ?

Reply:
I actually had something similar in mind initially, but gave up on it when I wasn't able to `min(len, buf.byteLength)` because `buf` could be a c pointer. Essentially, if `len` is given, it must not check `byteLength` because `buf` might be something unsafe, in turn making the entire function unsafe when specifying `buf` as an ArrayBuffer plus `len`.